### DISCUSSION TITLE
ULP via upython  & what language is 'gpio_dump_io_configuration()'?

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12869] [Libraries & Drivers] [answered]
The ULP repro shows how to embed ULP code in a upython program & have it assembled & run. In the examples some of the instructions (like #define DR_REG_RTCIO_BASE            0x3ff48400) look a bit C++ like to me while others (like move r1, state; st r0, r1, 0) are clearly ULP code.

Anyway I was reading https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gpio.html & saw gpio_dump_io_configuration() which won't run if I insert it into the ULP examples.
```
Traceback (most recent call last):
  File "<stdin>", line 38, in <module>
  File "/lib/esp32_ulp/__init__.py", line 19, in src_to_binary
  File "/lib/esp32_ulp/assemble.py", line 305, in assemble
  File "/lib/esp32_ulp/assemble.py", line 299, in assembler_pass
ValueError: Unknown opcode or directive: gpio_dump_io_configuration()
```
 Is this an espressiff IDE command? Is it a language (like upython or C+) or something else?

**Comment by jimmo** [comment, marked as answer]
It's assembly language (for the riscv-based cpu in the ULP). Like C, it also supports directives like `#define`, but it's assembly language.

The documentation you're looking at is the C api for the sdk for the main esp32 processor. You can't use this from rhe ULP.

You should look at 
https://github.com/micropython/micropython-esp32-ulp/blob/master/docs/index.rst

**Comment by kjm1102** [comment]
Huh. Wrong processor, wrong chip architecture. I need to improve my
situational awareness.
